packages/automerge-repo/src/storage/StorageSubsystem.ts

```diff
@@ -1,46 +1,148 @@
+import { keyHash, headsHash } from "./keyHash.js"
+import { chunkTypeFromKey } from "./chunkTypeFromKey.js"
+
+/**
+ * The storage subsystem is responsible for saving and loading Automerge documents to and from
+ * storage adapter. It also provides a generic key/value storage interface for other uses.
+ */
+export class StorageSubsystem {
+  /** Record of the latest heads we've loaded or saved for each document  */
+  #storedHeads: Map<DocumentId, A.Heads> = new Map()
 
-function keyHash(binary: Uint8Array) {
-  const hash = sha256.hash(binary)
-  const hashArray = Array.from(new Uint8Array(hash)) // convert buffer to byte array
-  const hashHex = hashArray.map(b => ("00" + b.toString(16)).slice(-2)).join("") // convert bytes to hex string
-  return hashHex
-}
+  /** Metadata on the chunks we've already loaded for each document */
+  #chunkInfos: Map<DocumentId, ChunkInfo[]> = new Map()
 
-function headsHash(heads: A.Heads): string {
-  const encoder = new TextEncoder()
-  const headsbinary = mergeArrays(heads.map((h: string) => encoder.encode(h)))
-  return keyHash(headsbinary)
-}
+  /** Flag to avoid compacting when a compaction is already underway */
+  #compacting = false
 
-export class StorageSubsystem {
-  #storageAdapter: StorageAdapter
-  #chunkInfos: Map<DocumentId, StorageChunkInfo[]> = new Map()
-  #storedHeads: Map<DocumentId, A.Heads> = new Map()
   #log = debug(`automerge-repo:storage-subsystem`)
 
-  #snapshotting = false
+  constructor(private storageAdapter: StorageAdapter) {}
+
+  // ARBITRARY KEY/VALUE STORAGE
+
+  // The `load`, `save`, and `remove` methods are for generic key/value storage, as opposed to
+  // Automerge documents. For example, they're used by the LocalFirstAuthProvider to persist the user
+  // and device information, as well as the encrypted team graph that encodes group membership and
+  // permissions.
+  //
+  // The namespace parameter is to prevent collisions with other users of the storage subsystem.
+  // Typically this will be the name of the plug-in, adapter, or other system that is using it. For
+  // example, the LocalFirstAuthProvider uses the namespace `LocalFirstAuthProvider`.
+
+  /** Loads a value from storage. */
+  async load(
+    /** Namespace to prevent collisions with other users of the storage subsystem. */
+    namespace: string,
+
+    /** Key to load. Typically a UUID or other unique identifier, but could be any string. */
+    key: string
+  ): Promise<Uint8Array | undefined> {
+    const storageKey = [namespace, key] as StorageKey
+    return await this.storageAdapter.load(storageKey)
+  }
+
+  /** Saves a value in storage. */
+  async save(
+    /** Namespace to prevent collisions with other users of the storage subsystem. */
+    namespace: string,
+
+    /** Key to load. Typically a UUID or other unique identifier, but could be any string. */
+    key: string,
 
-  constructor(storageAdapter: StorageAdapter) {
-    this.#storageAdapter = storageAdapter
```

Comment:
why isn't this private anymore?

Reply:
The `#` prefix marks class members as private but it's a JavaScript language feature rather than TypeScript so it's enforced at runtime.

Reply:
ah never mind, I responded on my phone & wasn't looking at the whole file. I've put that back the way it was.